### PR TITLE
Hotfix/746 multiple learning rate per layer patch

### PR DESCRIPTION
**Pull request by ybouilla**
**PR description**
For merging into `master` branch

Hotfix related to issue #746 : it introduces a more reliable way to send learning rates for scaffold (Fed-BioMed native implementation). Indeed, we claim to support pytorch models with different learning rates, but this was not working properly with non trainable layers such as batch norms. This patch should correct this issue, by instead of sending a list that contains all learning rates per layers, contains a mapping between layers having a learning rate and their values

**Developer Certificate Of Origin (DCO)**

By opening this merge request, you agree to the
[Developer Certificate of Origin (DCO)](https://github.com/fedbiomed/fedbiomed/blob/master/CONTRIBUTING.md#fed-biomed-developer-certificate-of-origin-dco)

This DCO essentially means that:

- you offer the changes under the same license agreement as the project, and
- you have the right to do that,
- you did not steal somebody else’s work.

**License**

Project code files should begin with these comment lines to help trace their origin:
```
# This file is originally part of Fed-BioMed
# SPDX-License-Identifier: Apache-2.0
```

Code files can be reused from another project with a compatible non-contaminating license.
They shall retain the original license and copyright mentions.
The `CREDIT.md` file and `credit/` directory shall be completed and updated accordingly.

**Guidelines for PR review**

General:

* give a glance to [DoD](https://fedbiomed.org/latest/developer/definition-of-done/)
* check [coding rules and coding style](https://fedbiomed.org/latest/developer/usage_and_tools/#coding-style)
* check docstrings (eg run `tests/docstrings/check_docstrings`)

Specific to some cases:

* update all conda envs consistently (`development` and `vpn`, Linux and MacOS)
* if modified researcher (eg new attributes in classes) check if breakpoint needs update (`breakpoint`/`load_breakpoint` in `Experiment()`, `save_state`/`load_state` in aggregators, strategies, secagg, etc.)
* if modified a component with versioning (config files, breakpoint, messaging protocol) then update the version following the rules in `common/utils/_versions.py`

### CODE REVIEW

File: fedbiomed/researcher/aggregators/scaffold.py [hunk 1]
```diff
@@ -244,18 +244,26 @@ def update_correction_states(
         # c_i^{t+1} = delta_i^t + (x^t - y_i^t) / (M * eta)
         for node_id, updates in model_updates.items():
             d_i = self.nodes_deltas[node_id]
-            self.nodes_states[node_id] = {
-                key: d_i[key] + val / (self.nodes_lr[node_id][idx] * n_updates)
-                for idx, (key, val) in enumerate(updates.items())
-            }
+            for (key, val) in updates.items():
+
+                if self.nodes_lr[node_id].get(key):
```
Comment (sharkovsky) [inline fedbiomed/researcher/aggregators/scaffold.py:249]:
```suggestion
                if self.nodes_lr[node_id].get(key) is not None:
```

File: fedbiomed/researcher/aggregators/scaffold.py [hunk 2]
```diff
@@ -244,18 +244,26 @@ def update_correction_states(
...
-            for key in self.global_state
-        }
+        for key in self.global_state:
+            self.global_state[key] = 0
+            for state in self.nodes_states.values():
+                if state.get(key) is not None:
+                    self.global_state[key] = (
+                        sum(state[key] for state in self.nodes_states.values())
+                            / len(self.nodes_states)
+
```
Comment (sharkovsky) [inline fedbiomed/researcher/aggregators/scaffold.py:263]:
```suggestion
```

File: fedbiomed/researcher/aggregators/scaffold.py [hunk 3]
```diff
@@ -244,18 +244,26 @@ def update_correction_states(
...
+                        )
+
+
```
Comment (sharkovsky) [inline fedbiomed/researcher/aggregators/scaffold.py:266]:
```suggestion
```

File: fedbiomed/researcher/aggregators/scaffold.py [hunk 4]
```diff
@@ -373,24 +381,19 @@ def set_nodes_learning_rate_after_training(
...
         n_model_layers = len(training_plan.get_model_params())
         for node_id in self._fds.node_ids():
-            lrs: List[float] = []
+            lrs: Dict[str, float] = {}
 
             if training_replies[n_round].get_index_from_node_id(node_id) is not None:
                 # get updated learning rate if provided...
                 node_idx: int = training_replies[n_round].get_index_from_node_id(node_id)
-                lrs += training_replies[n_round][node_idx]['optimizer_args'].get('lr')
+                lrs.update(training_replies[n_round][node_idx]['optimizer_args'].get('lr'))
```
Comment (sharkovsky) [inline fedbiomed/researcher/aggregators/scaffold.py:389]: is there ever a case where the two conditions below are both true?
1) `training_replies[n_round].get_index_from_node_id(node_id) is not None`
2) `training_replies[n_round][node_idx]['optimizer_args']` does NOT contain the `lr` key

If yes, then this will fail because `update` cannot be called with a `None` argument. 
But I'm not familiar enough with the whole process to know whether this scenario can happen or not.

File: fedbiomed/researcher/aggregators/scaffold.py [hunk 5]
```diff
@@ -373,24 +381,19 @@ def set_nodes_learning_rate_after_training(
...
-
-            if len(lrs) == 1:
-                # case where there is one learning rate
-                lr = lrs * n_model_layers
+                lrs.update(optim.get_learning_rate())
 
-            elif len(lrs) == n_model_layers:
-                # case where there are several learning rates value
+            if len(lrs) == n_model_layers:
                 lr = lrs
```
Comment (sharkovsky) [inline fedbiomed/researcher/aggregators/scaffold.py:398]:
```suggestion
```

File: fedbiomed/researcher/experiment.py [hunk 6]
```diff
@@ -1498,7 +1498,7 @@ def run_once(self, increase: bool = False, test_after: bool = False) -> int:
 
         # Ready to execute a training round using the job, strategy and aggregator
         if self._global_model is None:
-            self._global_model = self._job.training_plan.get_model_params()
+            self._global_model = self._job.training_plan.after_training_params()
```
Comment (sharkovsky) [inline fedbiomed/researcher/experiment.py:1501]: good catch!

**Review by sharkovsky — CHANGES_REQUESTED** [review]
Great work on this hotfix, my review is just a couple of questions/deletion of spurious blank line, and one request for changes in the docstrings.

File: fedbiomed/researcher/aggregators/scaffold.py (hunk 1 above)
Comment (ybouilla) [reply fedbiomed/researcher/aggregators/scaffold.py:249]: Yes you are right about this one!  if `self.nodes_lr[node_id].get(key)` returns 0, it should pass that statement

File: fedbiomed/common/optimizers/generic_optimizers.py [hunk 7]
```diff
@@ -235,11 +235,14 @@ def get_learning_rate(self) -> List[float]:
                 (as many as the number of the layers contained in the model)
```
[the remark this answers is not in the document]
Comment (ybouilla) [reply fedbiomed/common/optimizers/generic_optimizers.py:235]: Yes that is true, but to the best of my knowledge a `Dict` is a `Mapping`. Anyway, I will commit changes

**Review by sharkovsky — APPROVED** [review]
